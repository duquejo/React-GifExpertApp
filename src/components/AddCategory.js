import { useState } from "react";
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

  const [ inputValue, setInputValue] = useState(''); // ''

  const handleInputChange = ( e ) => {
    setInputValue( e.target.value ); // Setting Input event
  };

  const handleSubmit = ( e ) => {
    
    e.preventDefault();
    // Greater than 2 value

    if( inputValue.trim().length > 2 ) {
      // setCategories( [...categories, 'Fullmetal Alchemist'] );
      setCategories( cats => [ inputValue, ...cats ] ); // Callback with the last state
      setInputValue( '' );
    }
  };

  return (
    <form onSubmit={ handleSubmit }>
      <input 
        type="text"
        value={ inputValue }
        onChange={ handleInputChange }
        placeholder="Type something and press 'enter' for giphy search!"
      />
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
};