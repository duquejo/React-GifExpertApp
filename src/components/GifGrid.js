
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from 'prop-types';

export const GifGrid = ( { category } ) => {

  const { data:images, loading } = useFetchGifs( category );

  return (
    <>
      <h3 className="animate__animated animate__fadeIn"> { category } </h3>
      { loading && <p>Loading...</p> }
      <div className="card-grid">
        {
          images.map( img => <GifGridItem key={ img.id } { ...img } /> ) // Sending all properties with Object Rest operator
        }
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
};