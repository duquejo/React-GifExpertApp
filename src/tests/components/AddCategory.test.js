import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe('<AddCategory /> tests', () => {

  const setCategories = jest.fn(); // USing Jest Function
  const value = 'Hello world'; // Default value for input
  let wrapper;
  
  beforeEach( () => {
    jest.clearAllMocks(); // If we have any mock it will clear it.
    wrapper = shallow( <AddCategory setCategories={ setCategories }/> );
  });

  test('Should render <AddCategory /> successfully', () => {
    expect( wrapper ).toMatchSnapshot();
  });

  test('Should evaluate input text text changes', () => {

    const input = wrapper.find('input');

    input.simulate('change', { target: { value } } );

    const inputAfter = wrapper.find('input');
    expect( inputAfter.prop('value').trim() ).toBe( value );

  });
  
  test('should NOT post data without submit', () => {
    
    wrapper.find('form').simulate('submit', { preventDefault: () => {} } );
    expect( setCategories ).not.toHaveBeenCalled(); // Not called!
  });

  test('Should call setCategories and clear the text input', () => {
    
    const input = wrapper.find('input');
    
    input.simulate('change', { target: { value } } ); // Simulated input onChange
    wrapper.find('form').simulate('submit', { preventDefault: () => {} } ); // Simulate form onSubmit

    expect( setCategories ).toHaveBeenCalled(); // setCategories has been called 1 times
    expect( setCategories ).toHaveBeenCalledTimes(1);
    expect( setCategories ).toHaveBeenCalledWith( expect.any( Function ) ); // The setCategories has been called with any Function

    expect( input.prop('value').trim() ).toBe('');
    
  });
  
  
});