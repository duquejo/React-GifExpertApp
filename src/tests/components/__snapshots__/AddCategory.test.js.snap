// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCategory /> tests Should render <AddCategory /> successfully 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="Type something and press 'enter' for giphy search!"
    type="text"
    value=""
  />
</form>
`;
